lib/logdna/client.rb

```diff
@@ -1,164 +1,130 @@
+    def schedule_flush
+      @flush_scheduled = true
+      start_timer = lambda {
+        sleep(@exception_flag ? @retry_timeout : @flush_interval)
+        flush
+      }
+      thread = Thread.new { start_timer }
+      thread.join
     end
 
     def write_to_buffer(msg, opts)
-      return if msg.nil?
-      msg = encode_message(msg)
+      if @lock.try_lock
+        @buffer.concat(@side_messages) unless @side_messages.empty?
+        processed_message = process_message(msg, opts)
+        new_message_size = processed_message.to_s.bytesize
 
-      if @lock.locked?
-        @side_buffer.write(msg)
-        @side_messages.push(message_hash(msg, opts))
-        return
-      end
+        @buffer_byte_size += new_message_size
+        @buffer.push(processed_message)
 
-      check_side_buffer
-      buffer_size = @buffer.write(msg)
-      @messages.push(message_hash(msg, opts))
-      buffer_size
-    end
+        @lock.unlock if @lock.locked?
 
-    def queue_to_buffer(queue=@queue)
-      next_object = queue.shift
-      write_to_buffer(next_object[:msg], next_object[:opts])
-    end
+        flush if @flush_limit <= @buffer_byte_size
 
-    def process_buffer(buffer_size)
-      if buffer_size > @actual_byte_limit
-        @buffer_over_limit = true
-        flush()
-        @buffer_over_limit = false
+        schedule_flush unless @flush_scheduled
       else
-        @task = create_flush_task
+        @side_messages.push(process_message(msg, opts))
       end
     end
 
-    # this should be running synchronously if @buffer_over_limit i.e. called from self.buffer
-    # else asynchronously through @task
-    def flush()
-      if defined? @@request and !@@request.nil?
-        request_messages = []
-        @lock.synchronize do
-          request_messages = @messages
-          @buffer.truncate(0)
-          @messages = []
-        end
-        return if request_messages.empty?
+    def send_request
+      @request.body = {
+        e: "ls",
+        ls: @buffer.concat(@side_messages)
+      }.to_json
+      @side_messages.clear
 
-        real = {
-          e: 'ls',
-          ls: request_messages,
-        }.to_json
-
-        @@request.body = real
-        @response = Net::HTTP.start(@uri.hostname, @uri.port, use_ssl: @uri.scheme == 'https') do |http|
-          http.request(@@request)
+      begin
+        @response = Net::HTTP.start(
+          @uri.hostname,
+          @uri.port,
+          use_ssl: @uri.scheme == "https"
+        ) do |http|
+          http.request(@request)
         end
 
-        # don't kill @task if this was executed from self.buffer
-        # don't kill @task if there are queued messages
-        unless @buffer_over_limit || @messages.any? || @task.nil?
-          @task.shutdown
-          @task.kill
+        if @response.is_a?(Net::HTTPForbidden)
+          puts "Please provide a valid ingestion key"
+        elsif !@response.is_a?(Net::HTTPSuccess)
+          puts "The response is not successful "
```

Comment:
No retry in this case?

Reply:
no, there is no point in retrying if the ingestion is not valid or not provided

Reply:
But how do you know that that's why we didn't get a success response? What if there's some other error?

Reply:
I know it does not look like an ideal handling. But testing how this mechanism works in Ruby, I found this to be the best approach. Because Ruby does not really work well with general exception handling. I tested it with writing just rescue, and switching off the connection and providing a bad ingestion key.  The rescue part did not catch it. But it does catch if the specific exception names are defined.  I can probably call `handle_excpetion` lambda in the second (`!@response.is_a?(Net::HTTPSuccess)`)  condition.

Reply:
This isn't an exception though, it's a non-success response. I'd consider a retry in this case - there's no harm.